2024/incoming/ref-util.pod

```diff
@@ -0,0 +1,270 @@
+
+In the days of yore, you might have found yourself writing code like:
+
+    if ( ref $sleigh eq 'HASH' ) {
+        load_presents($sleigh);
+    }
+
+But alas! This is like finding coal in your stocking. 🎄 By using C<ref $sleigh eq 'HASH'>, we're peeking behind the curtains, exposing Perl's internal workings. We're grabbing the string representation of a reference and hoping it matches our expectations. If C<$sleigh> isn't a reference at all, C<ref> returns an empty string — leaving us out in the cold. (At least it's not C<undef>, which would likely wreak havoc during runtime).
+
+=head2 Unwrapping Ref::Util ✨
+
+Enter L<Ref::Util>, our festive hero that's been around for a while but might not have made it onto your wish list yet! With it, we can rewrite our code as:
+
+    use Ref::Util qw< is_hashref >;
+
+    if ( is_hashref($sleigh) ) {
+        load_presents($sleigh);
+    }
+
+Isn't that as refreshing as a winter's first snow? ❄️  The code is cleaner and more readable, but the true magic lies beneath the surface. L<Ref::Util> doesn't just sprinkle a bit of tinsel on your code — it B<supercharges> it!
+
+But wait, there's more under the tree! L<Ref::Util> provides a sleigh full of functions to check all sorts of references:
+
+=over 4
+
+=item *
+
+C<is_ref($variable)>
+
+=item *
+
+C<is_scalarref($variable)>
+
+=item *
+
+C<is_arrayref($variable)>
+
+=item *
+
+C<is_hashref($variable)>
+
+=item *
+
+C<is_coderef($variable)>
+
+=item *
+
+C<is_regexpref($variable)>
+
+=item *
+
+C<is_globref($variable)>
+
+=item *
+
+C<is_formatref($variable)>
+
+=item *
+
+C<is_ioref($variable)>
+
+=item *
+
+C<is_refref($variable)>
+
+=back
+
+And their C<is_plain_*> and C<is_blessed_*> counterparts.
+
+This comprehensive suite lets you check for any type of reference without exposing Perl's internals, making your code more robust and intention-revealing.
+
+Example:
+
+    use Ref::Util qw< is_blessed_arrayref >;
+
+    if ( is_blessed_arrayref($elves) ) {
+        assign_tasks($elves);
+    }
+
+Instead of:
+
+    use Scalar::Util qw< blessed reftype >;
+    if ( ref_type($elves) eq 'ARRAY' && blessed($elves) ) {
+        assign_tasks($elves);
+    }
+
+(You might consider calling C<blessed($elves) eq 'ARRAY'> but that doesn't differentiate between package name and reference type.)
+
+=head2 Performance: The Gift That Keeps on Giving 🎁
+
+Now, you might be thinking, "Sure, the code looks nicer, but is that all?" Oh, dear reader, the true magic lies in the B<massive speed improvements>! L<Ref::Util> doesn't just make your code more readable — it makes it B<faster>.
+
+Let's peek into Santa's workshop and look at some benchmarks:
+
+    ref():                                   5.335e+00 +/- 1.8e-02 (0.3%)
+    ref(), reftype(), !blessed():            1.5545e+01 +/- 3.1e-02 (0.2%)
+    Ref::Util::is_plain_arrayref (CustomOP): 2.7951e+00 +/- 6.2e-03 (0.2%)
+    Data::Util::is_array_ref:                5.9074e+00 +/- 7.5e-03 (0.1%)
```

Comment:
This is a bit of a personal peeve of mine, but I **hate** timings printed in scientific notation. I have to carefully check for the exponent first, note that the second entry has a different order of magnitude, and then look at the mantissa while keeping this in mind. Even worse for negative exponents! My brain is much better prepared to look at a list of decimal-point-aligned numbers. Additionally, these numbers should **always** be sorted in such a way that the **best** alternative comes up first!
```
Ref::Util::is_plain_arrayref (CustomOP):  2.7951 +/- 0.0062 (0.2%)
ref():                                    5.3350 +/- 0.0180 (0.3%)
Data::Util::is_array_ref:                 5.9074 +/- 0.0075 (0.1%)
ref(), reftype(), !blessed():            15.5450 +/- 0.0310 (0.2%)
```

Reply:
This is a good point. The version I had was just copy/pasted verbatim from the benchmarking tool. Your version is much more readable.

Reply:
Even better:
```
Ref::Util::is_plain_arrayref (CustomOP):  2.7951 +/- 0.0062 (0.2%) -- baseline
ref():                                    5.3350 +/- 0.0180 (0.3%) -- 1.90 times
Data::Util::is_array_ref:                 5.9074 +/- 0.0075 (0.1%) -- 2.11 times
ref(), reftype(), !blessed():            15.5450 +/- 0.0310 (0.2%) -- 5.56 times
```

Reply:
Added.